src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp

```diff
@@ -70,51 +73,60 @@ ov::OutputVector ov::pass::GroupQueryAttentionDecomposition::decompose(
+    const auto seqlens_elemi64 = register_new_node<v0::Convert>(seqlens_k, ov::element::i64);
+    const auto real_seqlens = register_new_node<v1::Add>(seqlens_elemi64, one);
 
     // Only consider batch is 1
-    auto seqlens_1d = register_new_node<v1::Reshape>(real_seqlens, one, false);
-    auto past_sequence_length = register_new_node<v1::Subtract>(seqlens_1d, current_sequence_length);
+    const auto seqlens_1d = register_new_node<v1::Reshape>(real_seqlens, one, false);
+    const auto past_seqlen = register_new_node<v1::Subtract>(seqlens_1d, current_seqlen);
+    const auto curr_seqlen_scalar = register_new_node<v0::Squeeze>(current_seqlen);
+
     if (do_rotary) {
-        Q = rotaryEmbedding(Q,
-                            past_sequence_length,
-                            seqlens_1d,
-                            cos_cache.get_node_shared_ptr(),
-                            sin_cache.get_node_shared_ptr(),
-                            head_size_node,
-                            rotary_interleaved);
-        K = rotaryEmbedding(K,
-                            past_sequence_length,
-                            seqlens_1d,
-                            cos_cache.get_node_shared_ptr(),
-                            sin_cache.get_node_shared_ptr(),
-                            head_size_node,
-                            rotary_interleaved);
+        ov::Output<ov::Node> position_ids =
+            register_new_node<v4::Range>(zero_without_shape, curr_seqlen_scalar, one_without_shape, ov::element::i64);
+        position_ids = register_new_node<v1::Add>(position_ids, past_seqlen);
+
+        const auto cos = register_new_node<v8::Gather>(cos_cache, position_ids, zero);
+        const auto sin = register_new_node<v8::Gather>(sin_cache, position_ids, zero);
+        Q = rotaryEmbedding(Q, cos, sin, rotary_interleaved);
+        K = rotaryEmbedding(K, cos, sin, rotary_interleaved);
     }
+    const auto is_static_input = K.get_partial_shape().is_static();
 
     auto construct_kv_cache = [&](const ov::Output<ov::Node>& past, const ov::Output<ov::Node>& current) {
-        auto past_datas = register_new_node<v8::Slice>(past, zero, past_sequence_length, one, two);
-        auto curr_datas = register_new_node<v8::Slice>(current, zero, current_sequence_length, one, two);
-        return register_new_node<v0::Concat>(ov::NodeVector{past_datas, curr_datas}, 2);
+        return register_new_node<v0::Concat>(ov::OutputVector{past, current}, 2);
     };
+    if (is_static_input) {
+        const auto current_kv_len_const = register_new_node(
+            v0::Constant::create(ov::element::i64, ov::Shape{1}, {K.get_partial_shape()[2].get_length()}));
+        const auto past_kv_len_const = register_new_node(
+            v0::Constant::create(ov::element::i64, ov::Shape{1}, {past_key.get_partial_shape()[2].get_length()}));
+        past_key = register_new_node<v8::Slice>(past_key, current_kv_len_const, past_kv_len_const, one, two);
```

Comment:
Do we have document which designs the cache layout for static shape ? 

From first glimpse, we may think the cache grows afterwards, which is 
```
index:
0->past_len->cur_len
data layout:
[past cache]|[current cache]
```
However, the code here assumes that past data is placed after current data, I think the memory growth direction is different from ordinary thinking. It's better that we could have a document or an agreement about this
```
index:
0->cur_len->past_len
data layout:
[current cache]|[past cache]
```

Reply:
Not correct, the latest cache always at the end of the buffer. This part want to pop the 0 at begin of the buffer. The L120 is the concat logic.

Reply:
Updated comments.